docs/rules/guard-for-in.md

```diff
@@ -10,6 +10,8 @@ for (key in foo) {
 
 Note that simply checking `foo.hasOwnProperty(key)` is likely to cause an error in some cases; see [no-prototype-builtins](no-prototype-builtins.md).
 
+As an alternative, you can use the [no-extend-native](no-extend-native.md) rule to prevent the `Object` prototype from being modified. This will ensure that it's safe to use a `for in` on bare objects, but you will still have to be careful using `for in` on other objects.
+
```

Comment:
If you will still have to be careful in some cases, then it isn't an alternative?

Reply:
Well, you have to be careful about something else: that you only use `for...in` on bare objects, rather than that you use the `hasOwnProperty` guard every time. It also has better performance and readability.

It's not, like, a strictly better alternative, but that's not what "alternative" means to me. To me, an "alternative" is just an option someone else might choose.

How would you phrase it?

Reply:
I'd expect from an alternative to be something that can be enforced with another rule or a combination of other rules, and make the rule be entirely redundant or entirely unnecessary.

For the `guard-for-in` rule, an alternative could be `no-extend-native` + some rule that would be able to detect and disallow `for...in` on an object that doesn't have `Object.prototype` or `null` as its prototype.

We could maybe put the note in "When Not To Use" section (some rules already have that section), as: "If you only use `for...in` on bare objects..." but I still have concerns because that constraint is not enforceable.

Reply:
I think putting it in the "When Not To Use" section makes sense. Others "When Not To Use" sections also mention unenforceable constraints.

Reply:
Thanks for the additional info!

My opinion on this is neutral, so if other team members think that it is a good idea to add this note to the "When Not To Use" section, I wouldn't be opposed.

Reply:
I think, at the very least, people _will_ turn it off (for instance, my codebase doesn't guard `for-in`, and I think this is very common in all codebases that make extensive use of `{}` rather than `Map`s), so a comment about how to mitigate the impact seems like it would be useful. At the very least, a link in the "See also" section.

Reply:
I'm also of the mind that this doesn't make sense to add to the documentation. The purpose of `guard-for-in` is to ensure you're aware that `for-in` is also including prototype properties. The `no-native-extend` rule is aimed for ensuring you don't modify the prototypes of existing classes. Ensuring people don't modify a prototype isn't the same as ensuring a `for-in` loop is aware of the prototype.

Reply:
`for-in` only includes iterable prototype properties, and native prototypes are basically universally non-iterable. `no-extend-native` fixes the problem for all normal use cases of `for-in` (to iterate `{}` being used as `Map` in environments too old to support `Map`).

Reply:
That is true provided that you aren’t importing a third-party module that modified a prototype, in which case the no-extend-native rule won’t catch the problem. It’s much safer to check each loop rather than trying to catch every place a prototype might have been modified.

Reply:
Yes, but developers don't always do the "much safer" option, and there are often reasons not to (such as performance limitations, readability, or carefully auditing every dependency – no remotely-common third-party module modifies the `Object` prototype). I don't think it's right to entirely ignore a safer option just because it's not the _safest_ option, just because a developer wants to choose a different trade-off.